pkg/autodiscovery/listeners/kube_services.go

```diff
@@ -225,14 +265,232 @@ func (l *KubeServiceListener) removeService(ksvc *v1.Service) {
 	if ok {
 		l.m.Lock()
 		delete(l.services, ksvc.UID)
+		delete(l.endpointsAnnotated, fmt.Sprintf("%s/%s", ksvc.Namespace, ksvc.Name))
 		l.m.Unlock()
 
 		l.delService <- svc
+
 	} else {
 		log.Debugf("Entity %s not found, not removing", ksvc.UID)
 	}
 }
 
+func (l *KubeServiceListener) addedEndpt(obj interface{}) {
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", obj)
+		return
+	}
+	l.createEndpoint(castedObj, false)
+}
+
+func (l *KubeServiceListener) deletedEndpt(obj interface{}) {
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoint type, got: %v", obj)
+		return
+	}
+	l.removeEndpoint(castedObj)
+}
+
+func (l *KubeServiceListener) updatedEndpt(old, obj interface{}) {
+	// Cast the updated object or return on failure
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", obj)
+		return
+	}
+	// Cast the old object, consider it an add on cast failure
+	castedOld, ok := old.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", old)
+		l.createEndpoint(castedObj, false)
+		return
+	}
+	if endpointsDiffer(castedObj, castedOld) {
+		l.removeEndpoint(castedObj)
+		l.createEndpoint(castedObj, false)
+	}
+}
+
+// endpointsDiffer compares two endpoints to only go forward
+// when relevant fields are changed. This logic must be
+// updated if more fields are used.
+func endpointsDiffer(first, second *v1.Endpoints) bool {
+	// Quick exit if resversion did not change
+	if first.ResourceVersion == second.ResourceVersion {
+		return false
+	}
+
+	// Subsets
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+func (l *KubeServiceListener) createEndpoint(kendpt *v1.Endpoints, firstRun bool) {
+	if kendpt == nil {
+		return
+	}
+	endptId := fmt.Sprintf("%s/%s", kendpt.Namespace, kendpt.Name)
+
+	l.m.Lock()
+	for id, isAnnotated := range l.endpointsAnnotated {
```

Comment:
Why are we ranging here instead of directly looking up `l.endpointsAnnotated[endpId]` ?

Also, we'll need to hold a RLock during the lookup